python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -241,40 +261,58 @@ def to_default_asset_events(
             is_generic_test = bool(attached_node_unique_id)
 
             if has_asset_def and is_asset_check and is_generic_test:
-                is_test_successful = node_status == TestStatus.Pass
-                severity = (
-                    AssetCheckSeverity.WARN
-                    if node_status == TestStatus.Warn
-                    else AssetCheckSeverity.ERROR
-                )
-
                 attached_node_resource_props: Dict[str, Any] = manifest["nodes"].get(
                     attached_node_unique_id
                 ) or manifest["sources"].get(attached_node_unique_id)
                 attached_node_asset_key = dagster_dbt_translator.get_asset_key(
                     attached_node_resource_props
                 )
+                check_name = event_node_info["node_name"]
+
+                if (
+                    context
+                    and AssetCheckKey(asset_key=attached_node_asset_key, name=check_name)
+                    in context.selected_asset_check_keys
+                ):
+                    is_test_successful = node_status == TestStatus.Pass
+                    severity = (
+                        AssetCheckSeverity.WARN
+                        if node_status == TestStatus.Warn
+                        else AssetCheckSeverity.ERROR
+                    )
 
-                yield AssetCheckResult(
-                    passed=is_test_successful,
-                    asset_key=attached_node_asset_key,
-                    check_name=event_node_info["node_name"],
-                    metadata=metadata,
-                    severity=severity,
-                )
-            else:
-                for upstream_unique_id in upstream_unique_ids:
-                    upstream_resource_props: Dict[str, Any] = manifest["nodes"].get(
-                        upstream_unique_id
-                    ) or manifest["sources"].get(upstream_unique_id)
-                    upstream_asset_key = dagster_dbt_translator.get_asset_key(
-                        upstream_resource_props
+                    yield AssetCheckResult(
+                        passed=is_test_successful,
+                        asset_key=attached_node_asset_key,
+                        check_name=check_name,
+                        metadata=metadata,
+                        severity=severity,
                     )
 
-                    yield AssetObservation(
-                        asset_key=upstream_asset_key,
+                # dbt's default indirect selection (eager) will select relationship tests
+                # on unselected assets, if they're compared with a selected asset.
+                # This doesn't match Dagster's default check selection which is to only
+                # select checks on selected assets. When we use eager, we may receive
+                # unexpected test results so we log those as observations as if
+                # asset checks were disabled.
+                else:
```

Comment:
this is a little controversial. Alternatives:
- set indirect selection to [cautious](https://docs.getdbt.com/reference/node-selection/test-selection-examples?indirect-selection-mode=cautious#cautious-mode) instead of leaving it up to the user (hence defaulting to eager). This matches dagster's test selection behavior, but I think it's odd that dagster would change this dbt default
- change the dagster default. I think we should do this (I'll file an issue), but it's a non trivial change through all the selection paths. I think this is a passable workaround in the meantime